emails/tests/mgmt_process_emails_from_sqs_tests.py

```diff
@@ -0,0 +1,301 @@
+    }
+
+
+@override_settings(
+    STATSD_ENABLED=True,
+    AWS_SQS_EMAIL_QUEUE_URL="https://sqs.us-east-1.amazonaws.example.com/111222333/queue-name",
+)
+def test_main_loop_metrics():
+    """The main loop can run with no messages to process."""
+    with MetricsMock() as mm:
+        res = Command(queue=fake_queue(), max_seconds=2).main_loop()
+    assert res["cycles"] == 1
+    mm.assert_gauge(
+        "fx.private.relay.email_queue_count", 1, ["queue:queue-name"])
+    mm.assert_gauge(
+        "fx.private.relay.email_queue_count_delayed", 2, ["queue:queue-name"]
+    )
+    mm.assert_gauge(
+        "fx.private.relay.email_queue_count_not_visible", 3, ["queue:queue-name"]
+    )
+
+
+def test_main_loop_one_message():
+    """The main loop can process a message."""
+    msg = fake_sqs_message(json.dumps(EMAIL_SNS_BODIES["s3_stored"]))
+    res = Command(queue=fake_queue([msg], []), max_seconds=3).main_loop()
+    assert res["total_messages"] == 1
+
+
+def test_main_loop_keyboard_interrupt():
+    """The main loop halts on Ctrl-C."""
+    queue = fake_queue()
+    res = Command(queue=fake_queue([], KeyboardInterrupt())).main_loop()
+    assert res["cycles"] == 1
+    assert res["exit_on"] == "interrupt"
+
+
+def test_main_loop_no_body():
+    """The main loop continues on a message without a JSON body."""
+    msg = fake_sqs_message("I am a string, not JSON")
+    res = Command(queue=fake_queue([msg], []), max_seconds=3).main_loop()
+    assert res["failed_messages"] == 1
+    assert res["cycles"] == 2
+    msg.delete.assert_not_called()
+
+
+def test_main_loop_no_body_deleted():
+    """The main loop deletes a message without a JSON body."""
+    msg = fake_sqs_message("I am a string, not JSON")
+    res = Command(
+        queue=fake_queue([msg], []), max_seconds=3, delete_failed_messages=True
+    ).main_loop()
+    assert res["failed_messages"] == 1
+    assert res["cycles"] == 2
+    msg.delete.assert_called_once_with()
+
+
+def test_main_loop_ses_temp_failure_retry(mock_sns_inbound_logic):
+    """The main loop retries a message with a temporary SES failure."""
+    temp_error = ClientError(
+        {
+            "Error": {
+                "Message": "Maximum sending rate exceeded.",
+                "Code": "ThrottlingException",
+            }
+        },
+        "",
+    )
+    mock_sns_inbound_logic.side_effect = (temp_error, None)
+    msg = fake_sqs_message(json.dumps(EMAIL_SNS_BODIES["s3_stored"]))
+    res = Command(queue=fake_queue([msg], []), max_seconds=4).main_loop()
+    assert res["total_messages"] == 1
+    assert res["pause_count"] == 1
+
+
+def test_main_loop_ses_temp_failure_twice(mock_sns_inbound_logic):
+    """A temporary error followed by a second error is not retried."""
+    temp_error = ClientError(
+        {
+            "Error": {
+                "Message": "Maximum sending rate exceeded.",
+                "Code": "ThrottlingException",
+            }
+        },
+        "",
+    )
```

Comment:
question (non-blocking): in `pytest`, could this be made into its own fixture that is requested in each of these 2 tests that needs it? to eliminate the code duplication?

Reply:
I created a function `make_client_error` to help construct the three `ClientErrors` tested, and switched the second one to test `AccountSendingPausedException`, which I'm guessing is the other temporary exception that has the phrase 'pause'

And while I'm writing this, I see one more ``ClientError`` that could use ``make_client_error``...